File: LeonCommonFunctionsDemo/LeonCommonFunctionsDemo/LeonTagsView/LeonTagsView.m
//
//  LeonTagsView.m
//  LeonCommonFunctionsDemo
//
//  Created by Leon on 2016/12/1.
//  Copyright © 2016年 Leon. All rights reserved.
//

#import "LeonTagsView.h"

@interface LeonTagsView ()

// 记录frame
@property (nonatomic, assign) CGRect lastFrame;
// 标签总高度
@property (nonatomic, assign) NSInteger totalHieght;

@property (nonatomic, strong) NSMutableArray *allBtnTagsArray;
@property (nonatomic, strong) NSMutableArray *selectedTagTitlesArray;

@end

@implementation LeonTagsView
//if you use code for your page
- (id)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self initViewAttributes];
        self.backgroundColor = KRANDOM_COLOR;
    }
    return self;
}
//if you use XIB for your page
- (id)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self initViewAttributes];
        NSLog(@"x = %f, y = %f, width = %f, height = %f", self.frame.origin.x, self.frame.origin.y, self.frame.size.width, self.frame.size.height);
        self.backgroundColor = KRANDOM_COLOR;
    }
    return self;
}

- (void)initViewAttributes {
    self.isKeepRight = NO;
    
    self.allBtnTagsArray = [NSMutableArray arrayWithCapacity:1];
    self.selectedTagTitlesArray = [NSMutableArray arrayWithCapacity:1];
    
    self.btnTextFont = [UIFont systemFontOfSize:17];
    self.btnCornerRadiusValue = 5;
    
    self.btnHorizontalPadding = 10;
    self.btnVerticalPadding = 5;
    
    self.btnHorizontalMargin = 10;
    self.btnVerticalMargin = 10;
    
    self.btnLeftMargin = 10.0;
    self.btnRightMargin = 10.0;
    
    self.btnNormalTextColor = [UIColor grayColor];
    self.btnSelectedTextColor = [UIColor redColor];
    self.btnDefaultTextColor = [UIColor whiteColor];
    
    self.btnNormalBackgroundColor = [UIColor cyanColor];
    self.btnSelectedBackgroundColor = [UIColor grayColor];
    self.btnDefaultBackgroundColor = [UIColor blueColor];
    
    self.btnNormalBorderColor = [UIColor yellowColor];
    self.btnSelectedBorderColor = [UIColor redColor];
    self.btnDefaultBorderColor = [UIColor cyanColor];
    
    self.leonTagsViewStyle = LeonTagsViewStyle_NoSelection;
}

- (void)setTagsArray:(NSArray *)array {
    //防止cell重用
    [self.subviews enumerateObjectsUsingBlock:^(__kindof UIView * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        UIView *view = obj;
        [view removeFromSuperview];
    }];
    
    __weak typeof(self) weakSelf = self;
    [weakSelf.allBtnTagsArray removeAllObjects];
    weakSelf.totalHieght = 0.0;
    
    CGRect lastBtnRect = CGRectZero;
    if (self.isKeepRight) {
        lastBtnRect.origin.x = self.bounds.size.width - self.btnRightMargin;
    } else {
        lastBtnRect.origin.x = self.btnLeftMargin;
    }
    weakSelf.lastFrame = lastBtnRect;
    [array enumerateObjectsUsingBlock:^(NSString *titleString, NSUInteger idx, BOOL * _Nonnull stop) {
        [self setTagBtnWithTagTitle:titleString tagID:@"-1"];
    }];
    
    [self setSomeTagBtnsSelected:self.selectedTagTitlesArray];
    
}

- (void)setTagsArray:(NSArray *)array defaultSelectedTagsArray:(NSArray *)defaultSelectedTags {
    [self setTagsArray:array];
    [self setSomeTagBtnsSelected: defaultSelectedTags];
}

- (void)setTagsArray:(NSArray *)array selectedTagsArray:(NSArray *)selectedTagsArray defaultTagsArray:(NSArray *)defaultArray {
    if (array.count) {
        [self setTagsArray:array];
    }
    if (selectedTagsArray.count) {
        [self setSomeTagBtnsSelected: selectedTagsArray];
    }
    if (defaultArray.count) {
        [self setDefaultTagBtnsAttributes:defaultArray];
    }
}

- (void)setTagBtnWithTagTitle:(NSString *)tagTitle tagID:(NSString *)tagID {
    UIButton *tagBtn = [UIButton buttonWithType:UIButtonTypeCustom];
    tagBtn.frame = CGRectZero;
    tagBtn.contentHorizontalAlignment = UIControlContentHorizontalAlignmentCenter;
    [tagBtn setTitle:tagTitle forState:UIControlStateNormal];
    tagBtn.titleLabel.font = self.btnTextFont;
    tagBtn.layer.cornerRadius = self.btnCornerRadiusValue;
    tagBtn.layer.masksToBounds = YES;
    [tagBtn addTarget:self action:@selector(tagBtnClickedAction:) forControlEvents:UIControlEventTouchUpInside];
    
    NSDictionary *attribute = @{NSFontAttributeName: self.btnTextFont};
    CGSize tagTitleSize = [tagTitle sizeWithAttributes: attribute];
    tagTitleSize.width += self.btnHorizontalPadding * 2;
    tagTitleSize.height += self.btnVerticalPadding * 2;
    CGRect tagNewRect = CGRectZero;
    if (self.isKeepRight) {
        if (self.lastFrame.origin.x - tagTitleSize.width - self.btnHorizontalMargin - self.btnLeftMargin < 0) {
            tagNewRect.origin = CGPointMake(self.bounds.size.width - self.btnRightMargin - tagTitleSize.width, self.lastFrame.origin.y + tagTitleSize.height + self.btnVerticalMargin);
        } else {
            CGFloat magin = self.lastFrame.size.width > 0 ? self.btnHorizontalMargin : 0;
            tagNewRect.origin = CGPointMake(self.lastFrame.origin.x - tagTitleSize.width - magin, self.lastFrame.origin.y);
        }
    } else {
        if (self.lastFrame.origin.x + self.lastFrame.size.width + tagTitleSize.width + self.btnHorizontalMargin > self.bounds.size.width - self.btnRightMargin) {
            tagNewRect.origin = CGPointMake(self.btnLeftMargin, self.lastFrame.origin.y + tagTitleSize.height + self.btnVerticalMargin);
        } else {
            CGFloat magin = self.lastFrame.size.width > 0 ? self.btnHorizontalMargin : 0;
            tagNewRect.origin = CGPointMake(self.lastFrame.origin.x + self.lastFrame.size.width + magin, self.lastFrame.origin.y);
        }
    }
    
    tagNewRect.size = tagTitleSize;
    tagBtn.frame = tagNewRect;
    self.lastFrame = tagBtn.frame;
    
    self.totalHieght = self.lastFrame.origin.y + self.lastFrame.size.height;
    [self setHight:self andHight:self.totalHieght];
    [self setBtnNormalState:tagBtn];
    [self.allBtnTagsArray addObject:tagBtn];
    [self addSubview:tagBtn];
}

#pragma mark-改变控件高度
- (void)setHight:(UIView *)view andHight:(CGFloat)hight {
    CGRect tempFrame = view.frame;
    tempFrame.size.height = hight;
    view.frame = tempFrame;
}

#pragma mark - I need Method

- (void)setBtnSelectedState:(UIButton *)btn {
    btn.layer.borderColor = self.btnSelectedBorderColor.CGColor;
    btn.layer.borderWidth = 1;
    [btn  setTitleColor:self.btnSelectedTextColor forState:UIControlStateNormal];
    btn.backgroundColor = self.btnSelectedBackgroundColor;
    btn.selected = YES;
    btn.userInteractionEnabled = YES;
}

- (void)setBtnNormalState:(UIButton *)btn {
    btn.layer.borderColor = self.btnNormalBorderColor.CGColor;
    btn.layer.borderWidth = 1;
    [btn  setTitleColor:self.btnNormalTextColor forState:UIControlStateNormal];
    btn.backgroundColor = self.btnNormalBackgroundColor;
    btn.selected = NO;
    btn.userInteractionEnabled = YES;
}

- (void)setBtnDefaultState:(UIButton *)btn {
    btn.layer.borderColor = self.btnDefaultBorderColor.CGColor;
    btn.layer.borderWidth = 1;
    [btn  setTitleColor:self.btnDefaultTextColor forState:UIControlStateNormal];
    btn.backgroundColor = self.btnDefaultBackgroundColor;
    btn.selected = NO;
    btn.userInteractionEnabled = NO;
}

- (void)setSomeTagBtnsSelected:(NSArray *)tagTitles {
    if (self.selectedTagTitlesArray.count) {
        tagTitles = self.selectedTagTitlesArray;
    }
    [self.allBtnTagsArray enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        UIButton *tagBtn = obj;
        if ([tagTitles containsObject: tagBtn.titleLabel.text]) {
            [self setBtnSelectedState: tagBtn];
            if (!self.selectedTagTitlesArray.count) {
                [self.selectedTagTitlesArray addObject:tagBtn.titleLabel.text];
            }
        }
    }];
}

- (void)setDefaultTagBtnsAttributes:(NSArray *)defaultTagsArray {
    [self.allBtnTagsArray enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        UIButton *tagBtn = obj;
        if ([defaultTagsArray containsObject: tagBtn.titleLabel.text]) {
            [self setBtnDefaultState:tagBtn];
        }
    }];
}

- (void)tagBtnClickedAction:(UIButton *)btn {
    typeof(self) weak_self = self;
    if (_leonTagsViewStyle == LeonTagsViewStyle_NoSelection) {
        return;
    } else if (_leonTagsViewStyle == LeonTagsViewStyle_SingleSelection) {
        if ([self.selectedTagTitlesArray containsObject:btn.titleLabel.text]) {
            [self.selectedTagTitlesArray removeObject:btn.titleLabel.text];
            [self setBtnNormalState:btn];
        } else {
            [self.allBtnTagsArray enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
                UIButton *tagBtn = obj;
                if ([weak_self.selectedTagTitlesArray containsObject:tagBtn.titleLabel.text]) {
                    [weak_self setBtnNormalState:tagBtn];
                    [weak_self.selectedTagTitlesArray removeObject:tagBtn.titleLabel.text];
                }
            }];
            [self.selectedTagTitlesArray addObject:btn.titleLabel.text];
            [self setBtnSelectedState:btn];
        }
        if (_returnSelectedTagTitlesArrayBlock) {
            self.returnSelectedTagTitlesArrayBlock(weak_self.selectedTagTitlesArray);
        }
    } else if (_leonTagsViewStyle == LeonTagsViewStyle_MultiSelection) {
        if ([self.selectedTagTitlesArray containsObject:btn.titleLabel.text]) {
            [self.selectedTagTitlesArray removeObject:btn.titleLabel.text];
            [self setBtnNormalState:btn];
        } else {
            [self.selectedTagTitlesArray addObject:btn.titleLabel.text];
            [self setBtnSelectedState:btn];
        }
        if (_returnSelectedTagTitlesArrayBlock) {
            self.returnSelectedTagTitlesArrayBlock(weak_self.selectedTagTitlesArray);
        }
    } else {
        //不设置这个 ”leonTagsViewStyle” 这个属性的话，那么直接返回
        return;
    }
}

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

@end

File: LeonCommonFunctionsDemo/LeonCommonFunctionsDemo/LeonTagsView/Swift版本/Leon_TagsView.swift
//
//  Leon_TagsView.swift
//  LeonCommonFunctionsDemo
//
//  Created by Leon on 2017/8/16.
//  Copyright © 2017年 Leon_. All rights reserved.
//

import UIKit

enum LeonTagsViewStyle {
    case LeonTagsViewStyle_MultiSelection
    case LeonTagsViewStyle_SingleSelection
    case LeonTagsViewStyle_NoSelection
    case LeonTagsViewStyle_DefaultSelection
}

//定义闭包类型
typealias ReturnSelectedTagTitlesArray = ([String]) ->Void

class Leon_TagsView: UIView {
    
    private var lastFrame: CGRect = CGRect()    //记录frame
    var totalHeight: CGFloat = 0.0      //标签总高度
    private var allBtnTagsArray: [UIButton] = []
    private var selectedTagTitlesArray: [String] = []
    private var ReturnSelectedTagTitlesArrayBlock: ReturnSelectedTagTitlesArray?
    //btn靠左靠右显示,默认靠左
    var isKeepRight: Bool = false
    
    //btn上文字字体
    var btnTextFont: UIFont = .systemFont(ofSize: 17)
    
    //单个btn的圆角值
    var btnCornerRadiusValue: CGFloat = 5.0
    
    //单个btn 文字左右两边距离btn左右边界的值
    var btnHorizontalPadding: CGFloat = 10.0
    
    //单个btn 文字上下两边距离btn上下边界的值
    var btnVerticalPadding: CGFloat = 5.0
    
    //btn之间的距离 (左右)
    var btnHorizontalMargin: CGFloat = 10.0
    
    //btn之间的距离 (上下)
    var btnVerticalMargin: CGFloat = 10.0
    
    //一行中，第一个tag距离左边边界的距离
    var btnLeftMargin: CGFloat = 10.0
    
    //一行中，最后一个tag距离右边边界的最小距离
    var btnRightMargin: CGFloat = 10.0
    
    //设置单个btn上字体的颜色
    var btnNormalTextColor: UIColor = .gray     //正常状态下
    var btnSelectedTextColor: UIColor = .red    //选中状态下
    var btnDefaultTextColor: UIColor = .white   //默认btn状态
    
    //btn的背景颜色
    var btnNormalBackgroundColor: UIColor = .cyan       //正常状态下
    var btnSelectedBackgroundColor: UIColor = .gray     //选中状态下
    var btnDefaultBackgroundColor: UIColor = .blue      //默认btn状态
    
    //btn的边界颜色
    var btnNormalBorderColor: UIColor = .yellow     //正常状态下
    var btnSelectedBorderColor: UIColor = .red      //选中状态下
    var btnDefaultBorderColor: UIColor = .cyan      //默认btn状态
    
    //设置 “多选” “单选” “不能点击” 状态
    var leonTagsViewStyle: LeonTagsViewStyle = .LeonTagsViewStyle_NoSelection
    
    
    //闭包回掉方法
    func ReturnSelectedTagTitlesArrayMethod(ReturnSelectedTagTitlesArrayFunction: @escaping (_ attributeType :[String])-> Void){
        ReturnSelectedTagTitlesArrayBlock = ReturnSelectedTagTitlesArrayFunction
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.backgroundColor = UIColor.clear
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        
        fatalError("init(coder:) has not been implemented")
    }
    
    
    public func setTagsArray(array: [String]) {
        //防止cell重用
        for(index, view) in subviews.enumerated() {
            print("测试中...\(index) + \(view)")
            view.removeFromSuperview()
        }
        self.allBtnTagsArray.removeAll()
        self.totalHeight = 0.0
        var lastBtnRect: CGRect = CGRect(x: 0, y: 0, width: 0, height: 0)
        if isKeepRight {
            lastBtnRect = CGRect(x: self.bounds.size.width - self.btnRightMargin, y: 0, width: 0, height: 0)
        } else {
            lastBtnRect = CGRect(x: Int(self.btnLeftMargin), y: 0, width: 0, height: 0)
        }
        self.lastFrame = lastBtnRect
        
        for contentStr in array {
            setTagBtnWithTagTitle(tagTitle: contentStr, tagID: "-1")
        }
        self.setSomeTagBtnsSelected(tagTitles: self.selectedTagTitlesArray)
    }
    
    public func setTagsArray(tagsArray: [String], selectedTags: [String]) {
        self.setTagsArray(array: tagsArray)
        self.setSomeTagBtnsSelected(tagTitles: selectedTags)
    }
    
    public func setTagsArray(array: [String], selectedTagsArray: [String], defaultArray: [String]) {
        if array.count > 0 {
            self.setTagsArray(array: array)
        }
        if selectedTagsArray.count > 0 {
            self.setSomeTagBtnsSelected(tagTitles: selectedTagsArray)
        }
        if defaultArray.count > 0 {
            self.setDefaultTagBtnsAttributes(defaultTagsArray: defaultArray)
        }
    }
    
    private func setSomeTagBtnsSelected(tagTitles: [String]) {
        var tagTitleArray: [String] = tagTitles
        
        if self.selectedTagTitlesArray.count > 0 {
            tagTitleArray = self.selectedTagTitlesArray
        }
        
        for btn in self.allBtnTagsArray {
            if tagTitleArray.contains((btn.titleLabel?.text)!) {
                self.setBtnSelectedState(btn: btn)
                if self.selectedTagTitlesArray.count <= 0 {
                    self.selectedTagTitlesArray.append((btn.titleLabel?.text)!)
                }
            }
        }
    }
    
    private func setDefaultTagBtnsAttributes(defaultTagsArray: [String]) {
        
        for btn in self.allBtnTagsArray {
            if defaultTagsArray.contains((btn.titleLabel?.text)!) {
                self.setBtnDefaultState(btn: btn)
            }
        }
    }
    
    private func setTagBtnWithTagTitle(tagTitle: String, tagID: String) {
        let tagBtn: UIButton = UIButton(type: .custom)
        tagBtn.frame = CGRect(x: 0, y: 0, width: 0, height: 0)
        tagBtn.contentHorizontalAlignment = .center
        tagBtn.setTitle(tagTitle as String, for: .normal)
        tagBtn.titleLabel?.font = self.btnTextFont
        tagBtn.layer.cornerRadius = CGFloat(self.btnCornerRadiusValue)
        tagBtn.layer.masksToBounds = true
        tagBtn.addTarget(self, action: #selector(tagBtnClickedAction(btn:)), for: .touchUpInside)
        
        let attribute = [NSFontAttributeName : self.btnTextFont]
        var tagTitleSize: CGSize = (tagTitle as NSString).size(attributes: attribute)
        tagTitleSize.width += CGFloat(self.btnHorizontalPadding * 2)
        tagTitleSize.height += CGFloat(self.btnVerticalPadding * 2)
        
        var tagNewRect: CGRect = CGRect(x: 0, y: 0, width: 0, height: 0)
        
        if isKeepRight {
            if (self.lastFrame.origin.x - tagTitleSize.width - self.btnHorizontalMargin - self.btnLeftMargin < 0) {
                tagNewRect.origin = CGPoint(x: self.bounds.size.width - self.btnRightMargin - tagTitleSize.width, y: self.lastFrame.origin.y + tagTitleSize.height + self.btnVerticalMargin)
            } else {
                let magin: CGFloat = self.lastFrame.size.width > 0 ? self.btnHorizontalMargin : 0
                tagNewRect.origin = CGPoint(x: self.lastFrame.origin.x - tagTitleSize.width - magin, y: self.lastFrame.origin.y)
            }
        } else {
            if CGFloat(self.lastFrame.origin.x) + CGFloat(self.lastFrame.size.width) + CGFloat(tagTitleSize.width) + CGFloat(self.btnHorizontalMargin) > CGFloat(self.bounds.size.width) - CGFloat(self.btnRightMargin) {
                tagNewRect.origin = CGPoint(x: CGFloat(self.btnLeftMargin), y: CGFloat(self.lastFrame.origin.y) + CGFloat(tagTitleSize.height) + CGFloat(self.btnVerticalMargin))
            } else {
                let magin: CGFloat = self.lastFrame.size.width > 0 ? self.btnHorizontalMargin : 0
                tagNewRect.origin = CGPoint(x: CGFloat(self.lastFrame.origin.x) + CGFloat(self.lastFrame.size.width) + magin, y: CGFloat(self.lastFrame.origin.y))
            }
        }
        
        tagNewRect.size = tagTitleSize
        tagBtn.frame = tagNewRect
        self.lastFrame = tagBtn.frame
        
        self.totalHeight = self.lastFrame.origin.y + self.lastFrame.height
        setHeight(view: self, height: self.totalHeight)
        setBtnNormalState(btn: tagBtn)
        self.allBtnTagsArray.append(tagBtn)
        self.addSubview(tagBtn)
    }
    
    
    @objc private func tagBtnClickedAction(btn: UIButton) {
        if self.leonTagsViewStyle == .LeonTagsViewStyle_NoSelection {
            return
        } else if self.leonTagsViewStyle == .LeonTagsViewStyle_SingleSelection {
            if self.selectedTagTitlesArray.contains((btn.titleLabel?.text)!) {
                
                let index = self.selectedTagTitlesArray.index(of: (btn.titleLabel?.text)!)
                self.selectedTagTitlesArray.remove(at: index!)
                setBtnNormalState(btn: btn)
            } else {
                for btn in self.allBtnTagsArray {
                    if self.selectedTagTitlesArray.contains((btn.titleLabel?.text)!) {
                        
                        let index = self.selectedTagTitlesArray.index(of: (btn.titleLabel?.text)!)
                        self.selectedTagTitlesArray.remove(at: index!)
                        setBtnNormalState(btn: btn)
                    }
                }
                self.selectedTagTitlesArray.append((btn.titleLabel?.text)!)
                self.setBtnSelectedState(btn: btn)
                if self.ReturnSelectedTagTitlesArrayBlock != nil{
                    self.ReturnSelectedTagTitlesArrayBlock!(self.selectedTagTitlesArray)
                }
            }
        } else if self.leonTagsViewStyle == .LeonTagsViewStyle_MultiSelection {
            if self.selectedTagTitlesArray.contains((btn.titleLabel?.text)!) {
                let index = self.selectedTagTitlesArray.index(of: (btn.titleLabel?.text)!)
                self.selectedTagTitlesArray.remove(at: index!)
                self.setBtnNormalState(btn: btn)
            } else {
                self.selectedTagTitlesArray.append((btn.titleLabel?.text)!)
                self.setBtnSelectedState(btn: btn)
            }
            if self.ReturnSelectedTagTitlesArrayBlock != nil {
                self.ReturnSelectedTagTitlesArrayBlock!(self.selectedTagTitlesArray)
            }
        } else {
            return
        }
        
    }
    //改变控件高度
    private func setHeight(view: UIView, height: CGFloat) {
        var tempFrame: CGRect = view.frame
        tempFrame.size.height = height
        view.frame = tempFrame
    }
    //设置btn正常展示属性
    private func setBtnNormalState(btn: UIButton) {
        btn.layer.borderColor = self.btnNormalBorderColor.cgColor
        btn.layer.borderWidth = 1
        btn.setTitleColor(self.btnNormalTextColor, for: .normal)
        btn.backgroundColor = self.btnNormalBackgroundColor
        btn.isUserInteractionEnabled = true
    }
    //设置btn选中btn属性
    private func setBtnSelectedState(btn: UIButton) {
        btn.layer.borderColor = self.btnSelectedBorderColor.cgColor
        btn.layer.borderWidth = 1
        btn.setTitleColor(self.btnSelectedTextColor, for: .normal)
        btn.backgroundColor = self.btnSelectedBackgroundColor
        btn.isUserInteractionEnabled = true
    }
    //设置默认btn属性
    private func setBtnDefaultState(btn: UIButton) {
        btn.layer.borderColor = self.btnDefaultBorderColor.cgColor
        btn.layer.borderWidth = 1
        btn.setTitleColor(self.btnDefaultTextColor, for: .normal)
        btn.backgroundColor = self.btnDefaultBackgroundColor
        btn.isUserInteractionEnabled = false
    }
    
    /*
    // Only override draw() if you perform custom drawing.
    // An empty implementation adversely affects performance during animation.
    override func draw(_ rect: CGRect) {
        // Drawing code
    }
    */

}
